jerry-core/ecma/builtin-objects/ecma-builtin-global.cpp

```diff
@@ -1325,6 +1325,131 @@ ecma_builtin_global_object_escape (ecma_value_t this_arg __attr_unused___, /**<
+  ecma_string_t *input_string_p = ecma_get_string_from_value (string);
+  /* 2. */
+  lit_utf8_size_t input_size = ecma_string_get_size (input_string_p);
+
+  /* 3. */
+  MEM_DEFINE_LOCAL_ARRAY (input_start_p, input_size, lit_utf8_byte_t);
+  ssize_t sz = ecma_string_to_utf8_string (input_string_p, input_start_p, (ssize_t) (input_size));
+  JERRY_ASSERT (sz >= 0);
+
+  lit_utf8_iterator_t iterator = lit_utf8_iterator_create (input_start_p, input_size);
+  /* 4. */
+  lit_utf8_size_t output_length = 0;
+  /* The length of input string is always greater than output string
+   * so we re-use the input string buffer. */
+  lit_utf8_byte_t *output_char_p = input_start_p;
+
+  /* The state of parsing that tells us where we are in an escape pattern.
+   * 0    we are outside of pattern,
+   * 1    found '%', start of pattern,
+   * 2    found first hex digit of '%xy' pattern
+   * 3    found valid '%xy' pattern
+   * 4    found 'u', start of '%uwxyz' pattern
+   * 5-7  found hex digits of '%uwxyz' pattern
+   * 8    found valid '%uwxyz' pattern
+   */
+  uint8_t status = 0;
+  lit_code_point_t code_point = 0;
+  lit_code_point_t hex_digits = 0;
+  /* 5. */
+  while (!lit_utf8_iterator_is_eos (&iterator))
+  {
+    /* 6. */
+    ecma_char_t ch = lit_utf8_iterator_read_next (&iterator);
+
+    /* 7-8. */
+    if (status == 0 && ch == LIT_CHAR_PERCENT)
+    {
+      /* Found '%' char, start of escape sequence. */
+      status = 1;
+    }
+    /* 9-10. */
+    else if (status == 1 && ch == LIT_CHAR_LOWERCASE_U)
+    {
+      /* Found 'u' char after '%'. */
+      status = 4;
+    }
+    else if (status > 0 && lit_char_is_hex_digit (ch))
+    {
+      /* Found hexadecimal digit in escape sequence. */
+      hex_digits = hex_digits * 16 + lit_char_hex_to_int (ch);
+      status++;
+    }
+
+    /* 11-13. */
+    if (status == 8)
+    {
+      /* Found valid '%uwxyz' escape, decode. */
+      output_char_p -= 5;
+      output_length -= 5;
+      status = 0;
+      code_point = hex_digits;
+      hex_digits = 0;
+    }
+    /* 14-17. */
+    else if (status == 3)
+    {
+      /* Found valid '%xy' escape, decode. */
+      output_char_p -= 2;
+      output_length -= 2;
+      status = 0;
+      code_point = hex_digits;
+      hex_digits = 0;
+    }
+    /* 18. */
+    else
+    {
+      /* Copying unescaped character. */
+      code_point = ch;
+
+      /* Handle surrogate pairs. */
+      if (lit_is_code_unit_high_surrogate (ch) && !lit_utf8_iterator_is_eos (&iterator))
+      {
+        ecma_char_t next_ch = lit_utf8_iterator_peek_next (&iterator);
+
+        if (lit_is_code_unit_low_surrogate (next_ch))
+        {
+          code_point = lit_convert_surrogate_pair_to_code_point (ch, next_ch);
+          lit_utf8_iterator_incr (&iterator);
+        }
+      }
+    }
+
+    lit_utf8_size_t lit_size = lit_code_point_to_utf8 (code_point, output_char_p);
+    output_char_p += lit_size;
+    output_length += lit_size;
```

Comment:
Seems that the `output_length` is synchronized with `output_char_p`. In the case, we could remove `output_length` and calculate it as `output_char_p - input_start_p`.